types: convey location guarantee in return types

gthb opened the pull request:
What
---

When `tokenize` is called with `withLocation: true` (or without `withLocation`, since `true` is the default), make the return type convey that the `loc` property of returned tokens will not be `undefined`.

Add the same type parameter on `parse` although its return type is currently just `object` (so it's there if/when that return type gets fleshed out).

Type-parametrize `addTokenMeta` so that this token type information is not lost when `tokenize` results are passed through it.

How
---

Make the required type changes after jsdoc emits types, using a patch file, since jsdoc does not support type expressions, and [does not seem likely to do so in the future][1]. This lets us deliver this type improvement without the effort of switching tooling for generating the types.

The patch will hopefully not conflict very often with future changes in Fx, as these signatures will probably not change very often But if it does, that will not go unnoticed as the `build:types` run-script will fail. If/when that happens, one can just manually edit `dist/fx.d.ts` and regenerate the patch with `git diff dist/fx.d.ts > convey-loc-guarantee-in-types.patch`.

[1]: https://github.com/jsdoc/jsdoc/issues/1917

gthb committed:
types: convey location guarantee in return types

When `tokenize` is called with `withLocation: true` (or without `withLocation`,
since `true` is the default), make the return type convey that the `loc`
property of returned tokens will not be `undefined`.

Add the same type parameter on `parse` although its return type is currently
just `object` (so it's there if/when that return type gets fleshed out).

Type-parametrize `addTokenMeta` so that this token type information is not lost
when `tokenize` results are passed through it.

borgar commented:
Unfortunately, `dist/fx.d.ts` is a generated file so this would all be nuked next time `npm run build:types` is run.

gthb committed:
Workaround: apply loc type guarantee with patch

The types are generated by jsdoc, which does not support nontrivial type
expressions and [does not seem likely to do so in the future][1]. Using
different tooling would be a hassle.

So just apply these minimal type changes with a patch when building types.

If future changes cause the jsdoc type output to conflict with this patch,
then that will not go unnoticed as the build:types run-script will fail.
If/when that happens, one can just manually edit `dist/fx.d.ts` and regenerate
the patch with `git diff dist/fx.d.ts > convey-loc-guarantee-in-types.patch`.

[1]: https://github.com/jsdoc/jsdoc/issues/1917

gthb commented:
> Unfortunately, `dist/fx.d.ts` is a generated file so this would all be nuked next time `npm run build:types` is run.

Missed that! Updated this to make `npm run build:types` produce the types with these changes. That's done with a patch since jsdoc doesn't support type expressions.

Alternatively that patch could be applied on install in GRID-client if you prefer, see https://github.com/GRID-is/GRID-client/pull/8394 for that.

borgar commented:
This patch is not something I am comfortable maintaining. A consumer that needs this can work around this by declaring wrapper/proxy functions, which I something I may consider adding to Fx later.
